src/shared/utilities/childProcess.ts

```diff
@@ -45,138 +76,153 @@ export class ChildProcess {
+                reportError: err => errorHandler(err instanceof Error ? err : new Error(err)),
             }
-        })
 
-        // Emitted when streams are closed.
-        this.childProcess.once('close', (code, signal) => {
-            const result = this.makeResult(code)
-            this.processResult = result
+            if (timeout) {
+                if (timeout?.completed) {
+                    throw new Error('Timeout token was already completed.')
+                }
 
-            if (params.onClose) {
-                params.onClose(code, signal)
+                timeout.timer.catch(err => errorHandler(err, true))
             }
 
-            this.childProcess!.stdout?.removeAllListeners()
-            this.childProcess!.stderr?.removeAllListeners()
-            this.childProcess!.removeAllListeners()
-        })
-
-        // Emitted when process exits or terminates.
-        // https://nodejs.org/api/child_process.html#child_process_class_childprocess
-        // - If the process exited, `code` is the final exit code of the process, else null.
-        // - If the process terminated because of a signal, `signal` is the name of the signal, else null.
-        // - One of `code` or `signal` will always be non-null.
-        this.childProcess.once('exit', (code, signal) => {
-            const result = this.makeResult(typeof code !== 'number' ? -1 : code)
-            this.processResult = result
-
-            if (params.onExit) {
-                params.onExit(code, signal)
+            // Async.
+            // See also crossSpawn.spawnSync().
+            // Arguments are forwarded[1] to node `child_process` module, see its documentation[2].
+            // [1] https://github.com/moxystudio/node-cross-spawn/blob/master/index.js
+            // [2] https://nodejs.org/api/child_process.html
+            try {
+                this.childProcess = crossSpawn.spawn(this.command, args, mergedOptions.spawnOptions)
+            } catch (err) {
+                return errorHandler(err as Error)
             }
-        })
+
+            // Emitted whenever:
+            //  1. Process could not be spawned, or
+            //  2. Process could not be killed, or
+            //  3. Sending a message to the child process failed.
+            // https://nodejs.org/api/child_process.html#child_process_class_childprocess
+            // We also register error event handlers on the output/error streams in case a lower level library fails
+            this.childProcess.on('error', errorHandler)
+            this.childProcess.stdout?.on('error', errorHandler)
+            this.childProcess.stderr?.on('error', errorHandler)
+
+            this.childProcess.stdout?.on('data', (data: { toString(): string }) => {
+                if (mergedOptions.collect) {
+                    this.stdoutChunks.push(data.toString())
+                }
+
+                mergedOptions.onStdout?.call(paramsContext, data.toString())
```

Comment:
Can a lot of these `.bind` and `.call` calls just use arrow functions so you don't have to manually add the context?

Reply:
The context is being assigned to `this` on the callback. I guess it could just be an argument as well but I'm wanting to lean into more with using `this` as opposed to typing `XYZContext` as the first argument on a bunch of things. Not against arrow functions or anything, just playing around with different styles.